open_instruct/finetune.py

```diff
@@ -381,12 +432,33 @@ def main():
         )
 
     if args.model_name_or_path:
-        model = AutoModelForCausalLM.from_pretrained(
-            args.model_name_or_path,
-            from_tf=bool(".ckpt" in args.model_name_or_path),
-            config=config,
-            low_cpu_mem_usage=args.low_cpu_mem_usage,
-        )
+        if args.use_qlora:
+            bnb_config = BitsAndBytesConfig(
+                load_in_4bit=True,
+                bnb_4bit_use_double_quant=True,
+                bnb_4bit_quant_type="nf4",
+                bnb_4bit_compute_dtype=torch.bfloat16,
+            )
+            device_index = accelerator.process_index
+            device_map = {"": device_index} # force data-parallel training.
+            model = AutoModelForCausalLM.from_pretrained(
+                args.model_name_or_path,
+                from_tf=bool(".ckpt" in args.model_name_or_path),
+                config=config,
+                load_in_4bit=True,
+                quantization_config=bnb_config,
+                device_map=device_map,
+                torch_dtype=torch.bfloat16,
+            )
+        else:
+            device_index = accelerator.process_index
+            device_map = {"": device_index} # force data-parallel training.
```

Comment:
Ah, good point. Yes, this still forces data-parallel training. You have to do this unless you use deepspeed (since FSDP doesn't currently work with mixed requires_grad), but probably the default for non-qlora should be to use deepspeed. Let me fix this up.

Reply:
Yeah let's use deepspeed by default for non-qlora